src/rustzcash.rs

```diff
@@ -1019,3 +1021,493 @@ pub extern "system" fn librustzcash_sprout_verify(
+
+    // Accumulate the value commitment in the context. We do this to check internal consistency.
+    {
+        let mut tmp = value_commitment.clone();
+        tmp = tmp.negate(); // Outputs subtract from the total.
+        tmp = tmp.add(&unsafe { &*ctx }.bvk, &JUBJUB);
+
+        // Update the context
+        unsafe { &mut *ctx }.bvk = tmp;
+    }
+
+    // Write the value commitment to the caller
+    value_commitment
+        .write(&mut (unsafe { &mut *cv })[..])
+        .expect("should be able to serialize rcv");
+
+    true
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_sapling_spend_sig(
+    ask: *const [c_uchar; 32],
+    ar: *const [c_uchar; 32],
+    sighash: *const [c_uchar; 32],
+    result: *mut [c_uchar; 64],
+) -> bool {
+    // The caller provides the re-randomization of `ak`.
+    let ar = match Fs::from_repr(read_fs(&(unsafe { &*ar })[..])) {
+        Ok(p) => p,
+        Err(_) => return false,
+    };
+
+    // The caller provides `ask`, the spend authorizing key.
+    let ask = match redjubjub::PrivateKey::<Bls12>::read(&(unsafe { &*ask })[..]) {
+        Ok(p) => p,
+        Err(_) => return false,
+    };
+
+    // We compute `rsk`...
+    let rsk = ask.randomize(ar);
+
+    // We compute `rk` from there (needed for key prefixing)
+    let rk =
+        redjubjub::PublicKey::from_private(&rsk, FixedGenerators::SpendingKeyGenerator, &JUBJUB);
+
+    // Compute the signature's message for rk/spend_auth_sig
+    let mut data_to_be_signed = [0u8; 64];
+    rk.0
+        .write(&mut data_to_be_signed[0..32])
+        .expect("message buffer should be 32 bytes");
+    (&mut data_to_be_signed[32..64]).copy_from_slice(&(unsafe { &*sighash })[..]);
+
+    // Do the signing
+    let mut rng = OsRng::new().expect("should be able to construct RNG");
+    let sig = rsk.sign(
+        &data_to_be_signed,
+        &mut rng,
+        FixedGenerators::SpendingKeyGenerator,
+        &JUBJUB,
+    );
+
+    // Write out the signature
+    sig.write(&mut (unsafe { &mut *result })[..])
+        .expect("result should be 64 bytes");
+
+    true
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_sapling_binding_sig(
+    ctx: *const SaplingProvingContext,
+    value_balance: int64_t,
+    sighash: *const [c_uchar; 32],
+    result: *mut [c_uchar; 64],
+) -> bool {
+    // Grab the current `bsk` from the context
+    let bsk = redjubjub::PrivateKey::<Bls12>(unsafe { &*ctx }.bsk);
+
+    // Grab the `bvk` using DerivePublic.
+    let bvk = redjubjub::PublicKey::from_private(
+        &bsk,
+        FixedGenerators::ValueCommitmentRandomness,
+        &JUBJUB,
+    );
+
+    // In order to check internal consistency, let's use the accumulated value
+    // commitments (as the verifier would) and apply valuebalance to compare
+    // against our derived bvk.
+    {
+        // Compute value balance
+        let mut value_balance = match compute_value_balance(value_balance) {
```

Comment:
So now I'm confused about this `compute_value_balance` method
https://github.com/zcash/librustzcash/blob/master/src/rustzcash.rs#L755
- why is it negating when something is negative? Shouldn't the `value_balance` be negative when more money is going in to the pool?
- I thought we were calling the *value* `value_balance` but in this method you're using it for the curve point
- Seems this `compute_value_balance` should get it's parameter as `const`

Reply:
* It's defined as a signed integer in the spec, so it must be negated. It is also defined to be subtracted in the balance equation.
* It computes valueBalance in the exponent
* There's no `const` in Rust -- all variable binds are immutable by default.

Reply:
I know you like to use shadowing as much as possible ;) but I would consider using different var/method name for value_balance in exponent compared to the scalar itself.